src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -176,17 +176,22 @@ private IEnumerable<TypeInferenceInfo> GetTypesSimple(ExpressionSyntax expressio
 
             private IEnumerable<TypeInferenceInfo> InferTypeInArrowExpressionClause(ArrowExpressionClauseSyntax arrowClause)
             {
-                if (arrowClause.IsParentKind(SyntaxKind.PropertyDeclaration))
-                {
-                    return InferTypeInPropertyDeclaration(arrowClause.Parent as PropertyDeclarationSyntax);
-                }
+                var parentSymbol = SemanticModel.GetDeclaredSymbol(arrowClause.Parent, CancellationToken);
+                var symbolReturnType = GetSymbolReturnType();
+
+                return symbolReturnType != null 
+                    ? SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(symbolReturnType))
+                    : SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
 
-                if (arrowClause.Parent is BaseMethodDeclarationSyntax)
+                ITypeSymbol GetSymbolReturnType()
                 {
-                    return InferTypeInBaseMethodDeclaration(arrowClause.Parent as BaseMethodDeclarationSyntax);
+                    switch (parentSymbol)
+                    {
+                        case IPropertySymbol property: return property.Type;
+                        case IMethodSymbol method: return method.ReturnType;
+                        default: return null;
+                    }
                 }
```

Comment:
Since this function is only called once I'd prefer writign it more directly:

```
ISymbol symbolReturnType;
switch (parentSymbol)
 {
       case IPropertySymbol property: 
          symbolReturnType = property.Type;
          break;
       case IMethodSymbol method: 
          symbolReturnType = method.ReturnType;
          break;
       default: 
          return null;
}


```

Reply:
Sorry, it was my suggestion to use a function... I think it's more concise and has avoids duplication (of the assignment).

Reply:
It also separates the logic, which I think is nice regardless of how many times it is called.

Reply:
If this was a method it could be reused by the code that deals with block bodies and return statement. There still has to be some modification to that because of the bug with local functions. Reusing this might be a nice solution.

Reply:
Making it a method would be fine. As it is it it actually does not separate the logic and can't be reused since it's using `parentSymbol` directly.

Reply:
Yes I would prefer either a method or non-capturing local function instead